File: ports/php/src/Noncharacters.php
<?php

declare(strict_types=1);

namespace UnicodePhp;

/// Detection and enumeration of the 66 designated Unicode noncharacters per
/// UAX #44 §5.6 / Unicode Standard 17.0 §23.7.
///
/// Two categories:
///
///   - BMP block:  U+FDD0 .. U+FDEF                (32 codepoints)
///   - Plane ends: U+nnFFFE / U+nnFFFF for n=0..16 (34 codepoints)
///
/// Total: 66.
final class Noncharacters
{
    /// Whether `cp` is one of the 66 designated Unicode noncharacters.
    public static function isNoncharacter(int $cp): bool
    {
        if ($cp >= 0xFDD0 && $cp <= 0xFDEF) {
            return true;
        }
        if ($cp > 0x10FFFF) {
            return false;
        }
        $low16 = $cp & 0xFFFF;
        return $low16 === 0xFFFE || $low16 === 0xFFFF;
    }

    /// Enumerate the 66 noncharacters in ascending order.
    ///
    /// @return list<int>
    public static function all(): array
    {
        $out = [];
        for ($cp = 0xFDD0; $cp <= 0xFDEF; $cp++) {
            $out[] = $cp;
        }
        for ($n = 0; $n <= 16; $n++) {
            $out[] = $n * 0x10000 + 0xFFFE;
            $out[] = $n * 0x10000 + 0xFFFF;
        }
        return $out;
    }
}

File: ports/elixir/lib/unicode_security/noncharacters.ex
defmodule UnicodeSecurity.Noncharacters do
  @moduledoc """
  The 66 designated Unicode noncharacters per UAX #44 §5.6 / Unicode 17.0 §23.7.

  Two categories: the BMP block U+FDD0..U+FDEF (32 codepoints) and the plane ends
  U+nnFFFE / U+nnFFFF for n = 0..16 (34 codepoints). Noncharacters are reserved
  for internal use; conformant Unicode text must not contain them in interchange.
  """

  import Bitwise

  @doc "Whether `cp` is one of the 66 designated Unicode noncharacters."
  @spec noncharacter?(integer()) :: boolean()
  def noncharacter?(cp) do
    cond do
      cp >= 0xFDD0 and cp <= 0xFDEF ->
        true

      cp > 0x10FFFF ->
        false

      true ->
        low16 = band(cp, 0xFFFF)
        low16 == 0xFFFE or low16 == 0xFFFF
    end
  end

  @doc "Enumerate the 66 noncharacters in ascending order."
  @spec all_noncharacters() :: [non_neg_integer()]
  def all_noncharacters do
    bmp = Enum.to_list(0xFDD0..0xFDEF)
    planes = Enum.flat_map(0..16, fn n -> [n * 0x10000 + 0xFFFE, n * 0x10000 + 0xFFFF] end)
    bmp ++ planes
  end
end
